{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Analysis of single cell RNAseq and cell-surface antibody expression\n",
    "### Initial processing via PIPseeker v0.52 (Fluent) and matrix files uploaded for downstream analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(Seurat)\n",
    "library(tidyverse)\n",
    "library(progeny)\n",
    "library(fgsea)\n",
    "library(dorothea)\n",
    "library(ggplot2)\n",
    "library(openxlsx)\n",
    "library(HGNChelper)\n",
    "library(clustifyr)\n",
    "library(cowplot)\n",
    "library(corrplot)\n",
    "library(metap)\n",
    "library(pheatmap)\n",
    "library(CytoTRACE)\n",
    "library(colorspace)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in parse(text = x, srcfile = src): <text>:2:14: unexpected '['\n1: #Set Filepaths and Load Data\n2: samplename = [\n                ^\n",
     "output_type": "error",
     "traceback": [
      "Error in parse(text = x, srcfile = src): <text>:2:14: unexpected '['\n1: #Set Filepaths and Load Data\n2: samplename = [\n                ^\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "#Set Filepaths and Load Data\n",
    "samplename = [] #set sample name\n",
    "read_data <- function(samplename) { \n",
    "    filepathbarcodes <- paste0(\"/home/\", samplename, \"/barcodes_\", samplename, '_.tsv')\n",
    "    filepathfeatures <- paste0(\"/home/\", samplename, \"/genes_\", samplename, '_.tsv')\n",
    "    filepathmat <- paste0(\"/home/\", samplename, \"/matrix_\", samplename, '_.mtx')\n",
    "  \n",
    "    barcodes <- read.delim(filepathbarcodes, header=F)\n",
    "    features <- read.delim(filepathfeatures, header=F)\n",
    "    mat <- Matrix::readMM(filepathmat)\n",
    "  \n",
    "    colnames(mat) <- barcodes[,1]\n",
    "    rownames(mat) <- make.unique(features$V2)\n",
    "    patient.data <- mat\n",
    "    \n",
    "    seurat1 <- CreateSeuratObject(counts = patient.data, project = samplename)\n",
    "    \n",
    "    seurat1[['percent.mito']] <- PercentageFeatureSet(seurat1, pattern = 'MT-')\n",
    "    seurat1 <- subset(seurat1, subset = percent.mito < 15 & nFeature_RNA > 200 & nFeature_RNA < 2500)\n",
    "    \n",
    "    seurat_RNA <- subset(seurat1, features = rownames(seurat1)[! grepl('^ADT-', x = rownames(seurat1))])\n",
    "    seurat_ADT <- subset(seurat1, features = rownames(seurat1)[grepl('^ADT-', x = rownames(seurat1))])\n",
    "    \n",
    "    return(c(seurat_RNA, seurat_ADT))\n",
    "}\n",
    "\n",
    "for (s in sampesnames) {\n",
    "    paste(\"seurat\",\"_\",s) <- readdata(s)\n",
    "    paste(\"seurat\",\"_\",s,\"_\",\"RNA\") <- paste(\"seurat\",\"_\",s)[[1]]\n",
    "    paste(\"seurat\",\"_\",s,\"_\",\"ADT\") <- paste(\"seurat\",\"_\",s)[[2]]\n",
    "\n",
    "}\n",
    "\n",
    "#Integrate RNA via RPCA\n",
    "ifnb.list_RNA = list(seurat_RNA[,:])\n",
    "ifnb.list_ADT = list(seurat_ADT[,:])\n",
    "\n",
    "ifnb.list_RNA <- lapply(X = ifnb.list2, FUN = function(x) {\n",
    "    x <- NormalizeData(x, verbose = FALSE)\n",
    "    x <- FindVariableFeatures(x, selection.method = \"vst\", nfeatures = 2000)\n",
    "})\n",
    "\n",
    "#Integrate ADT data with RPCA, normalization method CLR\n",
    "ifnb.list_ADT <- lapply(X = ifnb.list_ADT, FUN = function(x) {\n",
    "    x <- NormalizeData(x, normalization.method = 'CLR', margin = 2)\n",
    "    x <- FindVariableFeatures(x, selection.method = \"vst\")\n",
    "})\n",
    "\n",
    "Patientfeatures <- SelectIntegrationFeatures(object.list = ifnb.list3)\n",
    "ifnb.list_ADT <- lapply(X = ifnb.list3, FUN = function(x) {\n",
    "    x <- ScaleData(x, features = Patientfeatures, verbose = FALSE)\n",
    "    x <- RunPCA(x, features = Patientfeatures, verbose = FALSE)\n",
    "})\n",
    "\n",
    "Patientanchors <- FindIntegrationAnchors(object.list = ifnb.list_ADT, reduction = \"rpca\", dims = 1:15)\n",
    "Patient_ALL_ADT <- IntegrateData(anchorset = Patientanchors, dims = 1:15)\n",
    "\n",
    "#Scale integrated datasets\n",
    "DefaultAssay(Patient_ALL_ADT) <- \"integrated\"\n",
    "Patient_ALL_ADT <- ScaleData(Patient_ALL_ADT, features=rownames(Patient_ALL_ADT))\n",
    "\n",
    "DefaultAssay(Patient_ALL_RNA) <- \"integrated\"\n",
    "Patient_ALL_RNA <- ScaleData(Patient_ALL_RNA, verbose = FALSE)\n",
    "\n",
    "\n",
    "#PCA, Clustering, and Generation of RNA-derived UMAP (Figure 1E)\n",
    "#Right\n",
    "DefaultAssay(Patient_ALL_RNA_ADT) <- \"integrated\"\n",
    "\n",
    "Patient_ALL_RNA_ADT <- RunPCA(Patient_ALL_RNA_ADT, npcs = 30, verbose = FALSE)\n",
    "Patient_ALL_RNA_ADT <- RunUMAP(Patient_ALL_RNA_ADT, reduction = \"pca\", dims = 1:30)\n",
    "Patient_ALL_RNA_ADT <- FindNeighbors(Patient_ALL_RNA_ADT, reduction = \"pca\", dims = 1:30)\n",
    "Patient_ALL_RNA_ADT <- FindClusters(Patient_ALL_RNA_ADT, resolution = 0.6)\n",
    "\n",
    "ADToverlayedonRNAUMAP <- FeaturePlot(Patient_ALL_RNA_ADT, features_ADT, cols = c(\"purple\", \"yellow\"), reduction = 'umap', label = TRUE, combine = TRUE)\n",
    "#Left\n",
    "Patient_ALL_RNA_ADT <- SetIdent(Patient_ALL_RNA_ADT, value = Patient_ALL_RNA_ADT@meta.data$new_clusters)\n",
    "\n",
    "cellsbytype <- DimPlot(Patient_ALL_RNA_ADT, group.by = \"ident\", label = TRUE) + ggtitle('Cells by Type') + NoAxes()\n",
    "cellsbypatient <- DimPlot(Patient_ALL_RNA_ADT, group.by = \"patient\") + ggtitle('Cells by Patient') + NoAxes(\n",
    "options(repr.plot.width=16, repr.plot.height=8)\n",
    "cellsbytype + cellsbypatient\n",
    "    \n",
    "#Differential Gene Expression (Figure 1F)\n",
    "DefaultAssay(Patient_ALL_RNA_ADT) <- \"RNA\"\n",
    "listofclusters <- c(\"Leukemia\", 'B', 'T', \"Plasma B\", 'Erythroid', 'NK', \n",
    "                    'Myeloid DC', 'pDC', 'CD16+ Monocytes') \n",
    "    \n",
    "for (i in 0:(length(listofclusters)-1))\n",
    "{\n",
    "    LIST.CLUSTERS.and.MARKERS[[i+1]] <- FindConservedMarkers(Patient_ALL_RNA_ADT, ident.1 = listofclusters[[i+1]], grouping.var = \"orig.ident\", only.pos = TRUE)\n",
    "}\n",
    "LIST.CLUSTERS.and.MARKERS\n",
    "    \n",
    "new$cluster <- factor(new$cluster, levels = listofclusters)\n",
    "new <- new[order(new$cluster), ]\n",
    "\n",
    "heatmapconservedmarkers <- DoHeatmap(subset(Patient_ALL_RNA_ADT, downsample = 5000),features = new$gene)\n",
    "    \n",
    "    \n",
    "#Define ADT Clusters and Overlay on RNA-derived UMAP (Figure 4A)\n",
    "DefaultAssay(Patient_ALL_RNA_ADT) <- \"ADTonly\"\n",
    "    \n",
    "Patient_ALL_RNA_ADT <- SetIdent(Patient_ALL_RNA_ADT, value = Patient_ALL_RNA_ADT@meta.data$ADT_integrated_nn_res.0.6) \n",
    "ADT <-  c(\"ADT-IgG1\", \"ADT-HLA-DR\", \"ADT-CD3\", \"ADT-CD4\", \"ADT-CD5\",\n",
    "          \"ADT-CD7\", \"ADT-CD10\", \"ADT-CD11b\", \"ADT-CD13\",\n",
    "          \"ADT-CD14\", \"ADT-CD19\", \"ADT-CD22\", \"ADT-CD30\", \"ADT-CD33\", \n",
    "          \"ADT-CD34\", \"ADT-CD45\", \"ADT-CD56\", \"ADT-CD64\", \"ADT-CD117\")\n",
    "    \n",
    "DotPlot(Patient_ALL_RNA_ADT, features = rev(features_ADT), cols = c(\"blue\", \"red\"), dot.scale = 8) + RotatedAxis()\n",
    "    \n",
    "Patient_ALL_RNA_ADT$ADTclusters <- Idents(Patient_ALL_RNA_ADT)\n",
    "Patient_ALL_RNA_ADT <- SetIdent(Patient_ALL_RNA_ADT, value = Patient_ALL_RNA_ADT@meta.data$ADTclusters)\n",
    "ADToverlayedonADTUMAP <- FeaturePlot(Patient_ALL_RNA_ADT, features_ADT,  reduction = 'adt.umap', label = TRUE, combine = TRUE)\n",
    "    \n",
    "#Differential Gene Expression of ADT Clusters (Figure 4B)\n",
    "DefaultAssay(Patient_ALL_RNA_ADT) <- \"RNA\"\n",
    "Patient_ALL_RNA_ADT <- SetIdent(Patient_ALL_RNA_ADT, value = Patient_ALL_RNA_ADT@meta.data$ADTclusters)\n",
    "cluster_markers_RNAADT <- FindAllMarkers(Patient_ALL_RNA_ADT, only.pos = T)\n",
    "cluster_markers_RNAADT_10 <- cluster_markers_RNAADT %>%\n",
    "  group_by(cluster) %>%\n",
    "  top_n(10, avg_log2FC)\n",
    "    \n",
    "    \n",
    "#Cytotrace Analysis (Figure 5A, B, C)\n",
    "#Downsampled to 3000 cells per leukemia population per patient\n",
    "Patient_ALL_RNA_ADT <- SetIdent(Patient_ALL_RNA_ADT, value = Patient_ALL_RNA_ADT@meta.data$newclusters)\n",
    "leukemiacluster <- subset(Patient_ALL_RNA_ADT, idents = 'Leukemia')\n",
    "    \n",
    "DefaultAssay(Patient_ALL_RNA_ADT) <- 'RNA'\n",
    "cyto <- as.matrix(GetAssayData(Patient_ALL_RNA_ADT, \"counts\"))\n",
    "results <- CytoTRACE(cyto)\n",
    "    \n",
    "cyto_UMAP <- Patient_ALL_RNA_ADTdownsampled@reductions[[\"umap\"]]@cell.embeddings\n",
    "plotCytoTRACE(\n",
    "    cyto_obj = results,\n",
    "    phenotype = NULL,\n",
    "    emb = cyto_UMAP\n",
    ")\n",
    "\n",
    "corrmatrix <- cor(cyto, ADT)\n",
    "corrplot(corrmatrix, method=\"color\", is.corr=FALSE,  \n",
    "         pch.cex = 0.9,insig = 'label_sig', pch.col = 'grey20',\n",
    "         col = diverging_hcl(100, palette = \"Blue-Red 2\"),\n",
    "         tl.col = \"black\", cl.cex = 0.9, cl.ratio = 0.1, cl.pos = 'b', cl.offest = 0.5, cl.length = 3)\n",
    "    \n",
    "    \n",
    "#Split by CytoTRACE score\n",
    "Patient_ALL_RNA_ADTdownsampled$cyto_clusters <- Patient_ALL_RNA_ADTdownsampled$newclusters\n",
    "metadata <- Patient_ALL_RNA_ADTdownsampled[[]]\n",
    "metadata$Primitiveclusters <- as.character(metadata$Primitiveclusters)  \n",
    "\n",
    "metadata[metadata$newclusters == \"low_cytotrace\" & metadata$Cytotrace > 0.95, \"cyto_clusters\"] <- \"high_cytotrace\"\n",
    "Patient_ALL_RNA_ADTdownsampled$cyto)clusters <- metadata$cyto_clusters\n",
    "\n",
    "findmarkers <- FindMarkers(Patient_ALL_RNA_ADTdownsampled, ident.1 = 'high_cytotrace', ident.2 = 'low_cytotrace',\n",
    "                          logfc.threshold = log(0.25))\n",
    "\n",
    "subset <- subset(Patient_ALL_RNA_ADTdownsampled, idents = c('high_cytotrace', 'low_cytotrace'))\n",
    "heatmapfindmarkers <- DoHeatmap(subset, features = row.names(findmarkers))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
