{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Analysis of Joint-Profiled Single Cell (SC) DNA and Cell-Surface Protein Data\n",
    "### All data was initially processed on an open-source pipeline: https://github.com/AbateLab/DAb-seq\n",
    "### Resulting HDF5 files and H5AD files were then imported for downstream analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import division\n",
    "import itertools\n",
    "import math\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "import seaborn as sns; sns.set(style=\"white\", color_codes=True)\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as mcol\n",
    "plt.rcParams['axes.unicode_minus'] = False\n",
    "from sklearn.manifold import TSNE\n",
    "from scipy.stats import spearmanr\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from scipy.cluster.hierarchy import dendrogram, linkage, fcluster, set_link_color_palette\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.decomposition import PCA\n",
    "import os\n",
    "from IPython.display import display, HTML\n",
    "from scipy.stats.mstats import gmean\n",
    "from matplotlib.ticker import NullFormatter\n",
    "from scipy.stats import binned_statistic\n",
    "from IPython.display import display, HTML, Image\n",
    "import h5py\n",
    "import copy\n",
    "import umap\n",
    "matplotlib.rc('font', family='Helvetica')\n",
    "import importlib\n",
    "import re \n",
    "from matplotlib.pyplot import cm\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "import sys\n",
    "from scipy.signal import find_peaks\n",
    "from scipy.optimize import curve_fit\n",
    "from scipy.stats import norm\n",
    "import scipy.stats as stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set thresholds for quality and total depth\n",
    "min_quality = 30\n",
    "min_total_depth = 10\n",
    "min_alt_depth = 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set antibodies used in DAb-seq analysis\n",
    "abs = ['CD10', 'CD117', 'CD11b', 'CD123', 'CD13', 'CD14', 'CD15', 'CD16','CD19', 'CD22', 'CD3', 'CD30', \n",
    "          'CD33', 'CD34', 'CD38', 'CD4', 'CD5', 'CD56', 'CD64', 'CD7', 'CD71','CD8']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load Genotype and Variant Information From Compressed H5 File\n",
    "#The below script separates the h5 file into separate genotype and variant matrices\n",
    "\n",
    "def load_genotypes(genotypes_path):\n",
    "\n",
    "    with h5py.File(genotypes_path, 'r') as f:\n",
    "         \n",
    "        cell_barcodes = copy.deepcopy([c.decode('utf8').split('.')[0] for c in f['CELL_BARCODES']])\n",
    "        variants = copy.deepcopy([v.decode('utf8') for v in f['VARIANTS']])\n",
    "        \n",
    "        genotypes = pd.DataFrame(np.transpose(f['GT']), index=cell_barcodes, columns=variants).sort_index()\n",
    "        genotypes.index.name = 'cell_barcode'\n",
    "        \n",
    "        quality = pd.DataFrame(np.transpose(f['GQ']), index=cell_barcodes, columns=variants).sort_index()\n",
    "        quality.index.name = 'cell_barcode'\n",
    "        \n",
    "        total_depth = pd.DataFrame(np.transpose(f['DP']), index=cell_barcodes, columns=variants).sort_index()\n",
    "        total_depth.index.name = 'cell_barcode'\n",
    "        \n",
    "        alt_depth = pd.DataFrame(np.transpose(f['AD']), index=cell_barcodes, columns=variants).sort_index()\n",
    "        alt_depth.index.name = 'cell_barcode'\n",
    "                      \n",
    "    return genotypes, quality, total_depth, alt_depth\n",
    "\n",
    "def load_variants(variants_file_path):\n",
    "    variant_info = pd.read_csv(variants_file_path, sep='\\t', header=0, index_col='Name', low_memory=False)\n",
    "    variant_info.index.name = 'variant' \n",
    "    \n",
    "    return variant_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Filter variant calls from genotyping data based on pre-set metrics\n",
    "\n",
    "def filter_variants(genotypes,\n",
    "                    alt_depth,\n",
    "                    total_depth,\n",
    "                    quality,\n",
    "                    min_alt_depth,\n",
    "                    min_total_depth,\n",
    "                    min_quality):\n",
    "\n",
    "    genotypes[total_depth < min_total_depth] = 3\n",
    "    genotypes[((genotypes == 1) | (genotypes == 2)) & (alt_depth < min_alt_depth)] = 3\n",
    "    genotypes[quality < min_quality] = 3\n",
    "    genotypes[genotypes.isnull()] = 3\n",
    "    \n",
    "    # Quality = 0 for FLT3-ITD implies WT\n",
    "    genotypes_flt3 = genotypes.loc[:, genotypes.filter(like='FLT3-ITD', axis=1).columns]\n",
    "    quality_flt3 = quality.loc[:, quality.filter(like='FLT3-ITD', axis=1).columns]\n",
    "    \n",
    "    genotypes_flt3[quality_flt3 == 0] = 0\n",
    "    genotypes.loc[:, genotypes.filter(like='FLT3-ITD', axis=1).columns] = genotypes_flt3\n",
    "\n",
    "    return genotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set experimental metadata and filepaths, load data\n",
    "genotypes_path = 'MPAL1.genotypes.hdf5'\n",
    "variant_info_path = 'MPAL1.variants.tsv'\n",
    "antibodies_path = 'MPAL1_Ab.adata.h5ad'\n",
    "\n",
    "\n",
    "# Load Genotyping Data and Variant Info\n",
    "genotypes, quality, total_depth, alt_depth = load_genotypes(genotypes_path)\n",
    "variant_info = load_variants(variant_info_path)\n",
    "\n",
    "\n",
    "# Apply Quality and Depth Filters to All Genotypes\n",
    "genotypes = filter_variants(genotypes, alt_depth, total_depth, quality, min_alt_depth, min_total_depth, min_quality)\n",
    "\n",
    "\n",
    "# Load and Format Antibody File\n",
    "adata = ad.read_h5ad(antibodies_path)\n",
    "num = file.split('_Ab.adata')[0][-1]\n",
    "adata.obs_names = [bc+'-'+num for bc in adata.obs_names]\n",
    "\n",
    "with open('MPAL1.cells.tsv') as f:\n",
    "    geno = [l.strip().split('\\t') for l in f]\n",
    "genes = geno[0][1:]\n",
    "barcodes = [l[0].replace('M', '-') for l in geno[1:]]\n",
    "geno_counts = np.array([l[1:] for l in geno[1:]]).astype(int)\n",
    "\n",
    "adata_g = ad.AnnData(X=geno_counts)\n",
    "adata_g.obs_names = barcodes\n",
    "adata_g.var_names = genes\n",
    "\n",
    "\n",
    "#Identify Joint Barcodes between Genotypes File and Ab File\n",
    "intersect = np.sort(list(set(barcodes) & set(adata.obs_names)))\n",
    "adata = adata[intersect]\n",
    "adata_g = adata_g[intersect]\n",
    "len(intersect)\n",
    "\n",
    "\n",
    "#Filtering Genotype Data based on pre-set metrics\n",
    "#In the below analysis, we only consider variants for which > 0.001 reads are mutated for a single cell\n",
    "min_fraction_mutated = 0.001\n",
    "min_fraction_genotyped = 0.50\n",
    "\n",
    "number_genotyped = (genotypes != 3).sum(axis=0)\n",
    "fraction_genotyped = pd.DataFrame(number_genotyped / genotypes.shape[0], columns=['fraction_genotyped'])\n",
    "number_mutated = ((genotypes == 1) | (genotypes == 2)).sum(axis=0)\n",
    "fraction_mutated = pd.DataFrame(number_mutated.divide(number_genotyped), columns=['fraction_mutated']).dropna()\n",
    "\n",
    "\n",
    "#Filter Genotype Data based on ClinVar annotation\n",
    "#Remove non-intronic variants, remove synonymous variants, include only pathogenic variants (also manually reviewed)\n",
    "variant_info_1 = variant_info[(variant_info[\"SnpEff_HGVS_p\"].notnull())]\n",
    "genotypes = genotypes[list(variant_info_1.index)]\n",
    "variant_info_2 = variant_info[variant_info[\"SnpEff_Annotation\"] != \"synonymous_variant\"]\n",
    "genotypes = genotypes[list(variant_info_2.index)]\n",
    "variant_info_3 = variant_info[variant_info[\"SnpEff_Annotation_Impact\"] == 'HIGH']\n",
    "genotypes = genotypes[list(variant_info_3.index)]\n",
    "\n",
    "\n",
    "#Call Genetic Clones\n",
    "clonal_set = list(genotypes.columns[np.apply_along_axis(lambda x: sum(x) > 2, 0, final_NGTs.values[:, 1:])])\n",
    "clone_cutoff = 10\n",
    "\n",
    "def assign_clones(c):\n",
    "    genotypes_clones = c.iloc[:, 1:].sum().sort_values(ascending=False).index\n",
    "    clone_colnames = ['Clone'] + list(genotypes_clones)\n",
    "    clone_df = pd.concat([y['Cell'], c.iloc[:, 1:].loc[:, genotypes_clones]], axis=1)\n",
    "    clone_df['Clone'] = clone_df[genotypes_clones].apply(lambda x: '_'.join(x.index[x.values.argsort()[::-1]] >= clone_cutoff), axis=1)\n",
    "    clone_df = clone_df.loc[:, clone_colnames]\n",
    "    return clone_df\n",
    "\n",
    "clones = [assign_clones(c) for c in clonal_set]\n",
    "\n",
    "\n",
    "#CLR Transformation of Antibody Data\n",
    "def clr_transform(andat_raw, add_pseudocount:bool=False, features:list=[]):\n",
    "        if add_pseudocount:\n",
    "            clr_data = np.log((1+andat_raw[:,features].X)/gmean(andat_raw[:,features].X+1, axis=1).reshape(-1,1))\n",
    "        else:\n",
    "            clr_data = np.log(andat_raw[:,features].X/gmean(andat_raw[:,features].X, axis=1).reshape(-1,1))\n",
    "        return alr_data, clr_data, ilr_data\n",
    "            \n",
    "adata.var['Ab_umi'] = adata.X.sum(axis=0)\n",
    "adata.obs['gene_count'] = adata_g.X.sum(axis=1)\n",
    "adata_g.var['amplicon_count'] = adata_g.X.sum(axis=0)\n",
    "adata_g.obs['umi_counts'] = adata.obs['umi_counts']\n",
    "adata_g.obs['raw_counts'] = adata.obs['raw_counts']\n",
    "adata_ab = ad.AnnData(adata[:,abs].X, obs=adata[:,abs].obs, var=adata[:,signal].var)\n",
    "\n",
    "ab_clr = clr_transform(adata_ab, ['IgG_count','umi_counts','raw_counts','gene_count'], \n",
    "                                         components=2, antibodies=signal)\n",
    "\n",
    "#Generate Antibody-Derived UMAP\n",
    "trans = umap.UMAP(n_neighbors=50, random_state=42).fit(ab_clr)\n",
    "trans_results = trans.embedding_\n",
    "\n",
    "plt.xlabel('UMAP 1', fontsize=12, labelpad=12)\n",
    "plt.ylabel('UMAP 2', fontsize=12, labelpad=12)\n",
    "plt.scatter(x=trans_results[:, 0],\n",
    "            y=trans_results[:, 1],\n",
    "            s=10)\n",
    "\n",
    "#Overlay individual antibodies on the antibody-derived UMAP\n",
    "fig = plt.figure(figsize=(8,8))\n",
    "for i in range(len(ab_clr.columnes)):\n",
    "    ax = fig.add_subplot(5, 5, i + 1)\n",
    "    ax.scatter(trans_results[:, 0],\n",
    "               trans_results[:, 1],\n",
    "               alpha=0.8,\n",
    "               c=ab_clr[ab_clr_columns[i]],\n",
    "               cmap='magma',\n",
    "               s=1)\n",
    "    plt.title(ab_clr.columns[i].split('_')[0], fontsize=11)\n",
    "    plt.xticks([])\n",
    "    plt.yticks([])\n",
    "plt.tight_layout()\n",
    "\n",
    "#Overlay mutations of interest on antibody-derived UMAP (Figure 2C, 2G)\n",
    "plt.xticks([])\n",
    "plt.yticks([])\n",
    "clones =[] #set clone of interest here\n",
    "\n",
    "hue_map = {0: 'Wildtype',\n",
    "           1: 'Het. Mutant',\n",
    "           2: 'Hom. Mutant',\n",
    "           3: 'No Call'}\n",
    "    \n",
    "color_map = {'Wildtype': 'lightgrey',\n",
    "             'Het. Mutant': 'lightblue',\n",
    "             'Hom. Mutatant': 'darkblue'}\n",
    "\n",
    "sns.scatterplot(x=trans_results[:, 0],\n",
    "                y=trans_results[:, 1], \n",
    "                hue=[hue_map[g] for g in genotypes[clone]],\n",
    "                palette=color_map,\n",
    "                hue_order=['Het. Mutant','Hom. Mutant','Wildtype'],\n",
    "                edgecolor=None)\n",
    "\n",
    "#Generate split-violin plots comparing mutant vs non-mutant populations (Figure 2D, 2H)\n",
    "antibody = [] #set antibody of interest\n",
    "sns.violinplot(data=ab_clr, x='all', y=ab_clr.columns[antibody], hue=clones,\n",
    "               split=True, palette=['lightgrey','red'])\n",
    "    \n",
    "#Define Immunophenotypic Subpopulations via Hierarchical Clustering and overlay on antibody-derived UMAP (Figure 2D, 2H)\n",
    "Z = linkage(trans_results, 'centroid')\n",
    "max_d = 2   \n",
    "\n",
    "dendrogram(Z, leaf_rotation=90.,leaf_font_size=8.,truncate_mode='level', p=4,max_d=max_d)\n",
    "plt.axhline(y=max_d, c='white', lw=1.5, linestyle='dashed') \n",
    "ab_clusters = fcluster(Z, max_d, criterion='distance')  \n",
    "\n",
    "ab_color = dict(zip(sorted(list(set(ab_clusters))),\n",
    "                    [plt.cm.get_cmap('jet')(x) for x in np.linspace(0, 1, len(set(ab_clusters)))]))\n",
    "c = [ab_color[c] for c in ab_clusters]\n",
    "plt.scatter(trans_results[:, 0],\n",
    "            trans_results[:, 1], \n",
    "            c=c,s=5,edgecolor=None)\n",
    "\n",
    "#All patients loaded and pre-processed, as outlined above\n",
    "#Calculate T-statistic of antibody signal between mutant vs non-mutant populations (Figure 2B)\n",
    "#For this figure, all patients were concatenated into a single dataframe, \"patients\" where 0 is WT and 1 is mutant\n",
    "ts = pd.DataFrame(index = Abs)\n",
    "\n",
    "for p in patients:\n",
    "    t = stats.ttest_ind(p[abs_clr][0],p[abs_clr][1])\n",
    "    ts[p]=t[0]\n",
    "\n",
    "sns.clustermap(ts, cmap= c, row_cluster = True, col_cluster = False, metric = 'centroid')\n",
    "\n",
    "#Phylogeny Analysis (Figure 3)\n",
    "#Separate models were constructed for individual patients to calculate t-statistic\n",
    "import statsmodels.formula.api as smf\n",
    "\n",
    "for a in abs_clr.columns:\n",
    "    dat_ab = pd.DataFrame({abs_clr: dat[a], 'clones': pd.Categorical(dat['clones'])})\n",
    "    dat_ab.columns = [a, 'clones']\n",
    "    dat_ab['clones'] = dat_ab_1['clones'].cat.codes\n",
    "    dat_ab['clones'] = dat_ab_1['Clone'].replace(to_replace={4: np.nan})\n",
    "    dat_ab = pd.get_dummies(dat_ab_1, columns=['clones'], prefix='clones', prefix_sep='_', drop_first=True)\n",
    "    dat_ab.columns = [col.replace('.', '_') for col in dat_ab_1.columns]\n",
    "\n",
    "    fit = smf.ols(formula='a ~ clones', data=dat_ab_1).fit()\n",
    "    t_statistic = fit.summary().tables[1][:,2]\n",
    "    p = fit.summary().tables[1][:,3]\n",
    "    \n",
    "#All models then concatenated into single dataframe, \"phylogeny\" (Figure 3A)\n",
    "phylogeny = phylogeny.sort_values(['t_statistic'], ascending = False)\n",
    "sns.barplot(data=phylogeny,y='t_statistic',x='ab_clr',palette='Spectral',ci='sd')\n",
    "sns.stripplot(data=phylogeny,y='t_statistic',x='ab_clr',color='black',edgecolor='gray')\n",
    "plt.ylabel('Maximum t-statistic for Each Patient')\n",
    "plt.xlabel('Cell-Surface Protein')\n",
    "\n",
    "#Individual patient violinplots (Figure 3B)\n",
    "top_abs = phylogeny.sort_values(['t_statistic'][:,0:4]\n",
    "                                \n",
    "for i in range(len(top_abs.columns)):\n",
    "    if p <= 0.05*phylogeny.shape(): \n",
    "        ax = fig.add_subplot(5, 5, i + 1)\n",
    "        sns.violinplot(x = \"clone\",y = top_abs.columns[i],data = phylogeny, palette=\"red\")\n",
    "        plt.xticks([]), plt.xlabel(\"\"), plt.legend([],[], frameon=False)\n",
    "    elif p <= -0.05*phylogeny.shape():\n",
    "        ax = fig.add_subplot(5, 5, i + 1)\n",
    "        sns.violinplot(x = \"clone\",y = top_abs.columns[i],data = phylogeny, palette=\"turquoise\")\n",
    "        plt.xticks([]), plt.xlabel(\"\"), plt.legend([],[], frameon=False)\n",
    "    else: \n",
    "        ax = fig.add_subplot(5, 5, i + 1)\n",
    "        sns.violinplot(x = \"clone\",y = top_abs.columns[i],data = phylogeny, palette=\"grey\")\n",
    "        plt.xticks([]), plt.xlabel(\"\"), plt.legend([],[], frameon=False)\n",
    "                                \n",
    "plt.tight_layout(pad=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
