# Single Cell DAb-seq and CITE-seq analyses of Mixed Phenotypic Acute Leukemia (MPAL)

### Why MPAL?

Mixed Phenotypic Acute Leukemia (MPAL) is a rare acute leukemia characterized by blasts with both lymphoid and myeloid immunophenotypic markers. Survival of patients with MPAL is poor and inferior to that of the more common acute lymphoid and myeloid leukemias (ALL and AML). Due to the relative rarity and heterogeneous nature of MPAL, optimal therapeutic strategies remain uncertain. Emerging data suggests that sub-classification of MPAL may be needed to facilitate therapeutic decision making5. However, the full immunophenotypic, genetic, and transcriptomic changes that determine risk stratification of this complex disease have not been elucidated. Until recently, the technology to simultaneously determine immunophenotypic, genetic, and transcriptomic heterogeneity in MPAL has not existed. MPAL, with its definitionally ‘mixed’ immunophenotype, is uniquely poised to benefit from multiomic single cell (SC) sequencing analysis, which can, for the first time, quantify the relationship between these cellular factors to better understand the biologic origin of MPAL and the mechanism of its poor prognosis. 

### Our Approach

We analyzed samples from 14 adult patients with newly diagnosed MPAL using two SC technologies in parallel: DAb-Seq (SC DNA plus protein sequencing) and CITE-seq (SC RNA plus protein sequencing). 

<img width="400" alt="image" src="https://user-images.githubusercontent.com/63078801/230311270-805c194a-3a2b-45f5-b483-8581bd5942c8.png">

Here, you will find code describing our approach for downstream analysis these SC DAb-seq and SC PIPseq data sets. All DAb-seq data was initially processed using an open-source pipeline, available [here,](https://github.com/AbateLab/DAb-seq) and all CITE-seq data was initiall processed using [PIPseeker,](https://www.fluentbio.com/products/pipseeker-software-for-data-analysis) developed by Fluent BioSciences. All data can be accessed through GEO. A pre-print of the manuscrupt is available on bioRxiv.


### Questions?

For correspondence regarding our manuscript, please contact [Cathy Smith](mailto:catherine.smith@ucsf.edu). Questions regarding wet-lab sample processing can be directed to [Cheryl Peretz](mailto:cheryl.peretz@ucsf.edu) and questions about bioinformatic analyses can be directed to [Vanessa Kennedy](mailto:vanessa.kennedy@ucsf.edu). 

